src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,546 @@
+                        initiator_static_ip, jbof_static_ip,
+                        jbof_static_ip_from_initiator_ip,
+                        static_ip_netmask_int, static_ip_netmask_str,
+                        static_mtu)
+
+
+class JBOFModel(sa.Model):
+    __tablename__ = 'storage_jbof'
+
+    id = sa.Column(sa.Integer(), primary_key=True)
+
+    jbof_description = sa.Column(sa.String(120), nullable=True)
+
+    # When performing static (code-based) assignment of data-plane IPs, we
+    # want each JBOD to have a deterministic unique index that counts up from
+    # zero (without any gaps, which rules out using the id).  This will not be
+    # part of the public API.
+    jbof_index = sa.Column(sa.Integer(), unique=True)
+    jbof_uuid = sa.Column(sa.Text(), nullable=False, unique=True)
+
+    jbof_mgmt_ip1 = sa.Column(sa.String(45), nullable=False)
+    jbof_mgmt_ip2 = sa.Column(sa.String(45))
+    jbof_mgmt_username = sa.Column(sa.String(120))
+    jbof_mgmt_password = sa.Column(sa.EncryptedText())
+
+
+class JBOFService(CRUDService):
+
+    class Config:
+        service = 'jbof'
+        datastore = 'storage.jbof'
+        datastore_prefix = "jbof_"
+        cli_private = True
+        role_prefix = 'JBOF'
+
+    ENTRY = Dict(
+        'jbof_entry',
+        Int('id', required=True),
+
+        Str('description'),
+
+        # Redfish
+        IPAddr('mgmt_ip1', required=True),
+        IPAddr('mgmt_ip2', required=False),
+        Str('mgmt_username', required=True),
+        Password('mgmt_password', required=True),
+    )
+
+    @private
+    async def add_index(self, data):
+        """Add a private unique index (0-255) to the entry if not already present."""
+        if 'index' not in data:
+            index = await self.middleware.call('jbof.next_index')
+            if index is not None:
+                data['index'] = index
+        return data
+
+    @private
+    async def validate(self, data, schema_name, old=None):
+        verrors = ValidationErrors()
+
+        # Check license
+        license_count = await self.middleware.call("jbof.licensed")
+        if license_count == 0:
+            verrors.add(f"{schema_name}.mgmt_ip1", "This feature is not licensed")
+        else:
+            if old is None:
+                # We're adding a new JBOF - have we exceeded the license?
+                count = await self.middleware.call('jbof.query', [], {'count': True})
+                if count >= license_count:
+                    verrors.add(f"{schema_name}.mgmt_ip1", f"Already configured the number of licensed emclosures: {license_count}")
+
+        # Ensure redfish connects to mgmt1 (incl login)
+        mgmt_ip1 = data.get('mgmt_ip1')
+        if not RedfishClient.is_redfish(mgmt_ip1):
+            verrors.add(f"{schema_name}.mgmt_ip1", "Not a redfish management interface")
+        else:
+            redfish1 = RedfishClient(f'https://{mgmt_ip1}')
```

Comment:
Is there any chance of having `is_redfish`, but subsequent request to get redfish root failing here?

Reply:
Logically it should work, but there _could_ suddenly be a network outage, etc.  I was going to allow the exception to bubble up in that case.  Do you think I should catch instead?

Reply:
No. I just wanted to make sure that's intentional.